src/pmdas/amdgpu/localdrm.c

```diff
@@ -0,0 +1,323 @@
+    return status;
+
+  /* The GPU speed is the memory clock (GFX_MCLK)
+   * The GDDRx memory speed is the shader clock (GFX_SCLK)
+   */  if (amdgpu_query_sensor_info(amdgpu_device, AMDGPU_INFO_SENSOR_GFX_SCLK, sizeof(*value), value) < 0)
+    return DRM_ERROR_NO_DATA;
+
+  releaseAMDDevice(amdgpu_device, fd);
+
+  return DRM_SUCCESS;
+}
+
+int localDRMDeviceGetGPUClock(drmDevicePtr device, uint32_t *value) {
+  amdgpu_device_handle amdgpu_device;
+  int fd = -1;
+  int status = DRM_SUCCESS;
+
+  if ((status = getAMDDevice(device, &amdgpu_device, &fd)) != DRM_SUCCESS)
+    return status;
+
+  /* The GPU speed is the memory clock (GFX_MCLK)
+   * The GDDRx memory speed is the shader clock (GFX_SCLK)
+   */
+  if (amdgpu_query_sensor_info(amdgpu_device, AMDGPU_INFO_SENSOR_GFX_MCLK, sizeof(*value), value) < 0)
+    return DRM_ERROR_NO_DATA;
+
+  releaseAMDDevice(amdgpu_device, fd);
+
+  return DRM_SUCCESS;
+}
+
+int localDRMDeviceGetTemperature(drmDevicePtr device, uint32_t *value) {
+  amdgpu_device_handle amdgpu_device;
+  int fd = -1;
+  int status = DRM_SUCCESS;
+
+  if ((status = getAMDDevice(device, &amdgpu_device, &fd)) != DRM_SUCCESS)
+    return status;
+
+  if (amdgpu_query_sensor_info(amdgpu_device, AMDGPU_INFO_SENSOR_GPU_TEMP, sizeof(*value), value) < 0)
+    return DRM_ERROR_NO_DATA;
+
+  releaseAMDDevice(amdgpu_device, fd);
+
+  return DRM_SUCCESS;
+}
+
+int localDRMDeviceGetGPULoad(drmDevicePtr device, uint32_t *value) {
+  amdgpu_device_handle amdgpu_device;
+  int fd = -1;
+  int status = DRM_SUCCESS;
+
+  if ((status = getAMDDevice(device, &amdgpu_device, &fd)) != DRM_SUCCESS)
+    return status;
+
+  if (amdgpu_query_sensor_info(amdgpu_device, AMDGPU_INFO_SENSOR_GPU_LOAD, sizeof(*value), value) < 0)
+    return DRM_ERROR_NO_DATA;
+
+  releaseAMDDevice(amdgpu_device, fd);
+
+  return DRM_SUCCESS;
+}
+
+int localDRMDeviceGetGPUAveragePower(drmDevicePtr device, uint32_t *value) {
+  amdgpu_device_handle amdgpu_device;
+  int fd = -1;
+  int status = DRM_SUCCESS;
+
+  if ((status = getAMDDevice(device, &amdgpu_device, &fd)) != DRM_SUCCESS)
+    return status;
+
+  if (amdgpu_query_sensor_info(amdgpu_device, AMDGPU_INFO_SENSOR_GPU_AVG_POWER, sizeof(*value), value) < 0)
+    return DRM_ERROR_NO_DATA;
+
+  releaseAMDDevice(amdgpu_device, fd);
+
+  return DRM_SUCCESS;
+}
+
+int localDRMDeviceGetMemoryInfo(drmDevicePtr device, drmMemory_t *memory) {
+  amdgpu_device_handle amdgpu_device;
+  struct drm_amdgpu_memory_info mem;
+  int fd = -1;
+  int status = DRM_SUCCESS;
+
+  if ((status = getAMDDevice(device, &amdgpu_device, &fd)) != DRM_SUCCESS)
+    return status;
+
+  if (amdgpu_query_info(amdgpu_device, AMDGPU_INFO_MEMORY, sizeof(mem), &mem) < 0)
+    return DRM_ERROR_NO_DATA;
+
+  memory->total = mem.vram.total_heap_size;
+  memory->usable = mem.vram.usable_heap_size;
+  memory->used = mem.vram.heap_usage;
+
+  releaseAMDDevice(amdgpu_device, fd);
+
+  return DRM_SUCCESS;
+}
+
```

Comment:
There's alot of opening and closing of device files here, with the way these interfaces are setup (I think this is a bit based on nvidia again, and thats maybe made it more complex than necessary).

Going back to the earlier comment about need_refresh - currently we call all of these APIs on every fetch, for every GPU.  If we are going to do that, we can collapse all of these library calls into a single function that opens and closes each GPU fd a maximum of once per fetch, and avoid system time overheads associated with doing this "for every GPU for every metric" (i.e. ~10x less work if we have 10 metrics).